cli_parser: man-pages and help text generation using Homebrew::CLI::Parser

GauthamGoli opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
https://github.com/Homebrew/brew/pull/4271#issuecomment-394177933
This PR is about generating `--help` text using `OptionParser` and to get feedback on the interface, and the way help text is being shown as its different from what we currently show.

`brew audit --help` would return something like this.

```
Usage: brew audit [options] [<formulae>]

Check <formulae> for Homebrew coding style violations. This should be
run before submitting a new formula.

If no <formulae> are provided, all of them are checked.
        --strict                     Run additional style checks, including Rubocop style checks.
        --online                     Run additional slower style checks that require a
                                     network connection.
        --new-formula                Run various additional style checks to determine if a new formula
                                     is eligible for Homebrew. This should be used when creating
                                     new formula and implies --strict and --online.
        --fix                        Fix style violations automatically using
                                     RuboCop's auto-correct feature.
        --display-cop-names          Include the RuboCop cop name for each violation
                                     in the output.
        --display-filename           Prefix everyline of output with name of the file or
                                     formula being audited, to make output easy to grep.
    -D, --audit-debug                Activates debugging and profiling
        --only                       Passing --only=method will run only the methods named audit_method,
                                     `method` should be a comma-separated list.
        --except                     Passing --except=method will run only the methods named audit_method,
                                     `method` should be a comma-separated list.
        --only-cops                  Passing --only-cops=cops will check for violations of only the listed
                                     RuboCop cops. `cops` should be a comma-separated list of cop names.
        --except-cops                Passing --except-cops=cops will skip checking the listed
                                     RuboCop cops violations. `cops` should be a comma-separated list of cop names.
    -v, --verbose                    Verbose mode.
    -d, --debug                      Display debug info.
    -h, --help                       Show this message
```

MikeMcQuaid commented:
Can you run `brew man` so we can see the different help text output? Thanks!

GauthamGoli commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -54,19 +54,27 @@ module Homebrew
+
+        If no <formulae> are provided, all of them are checked.
+      EOS
+      switch      "--strict", description: "Run additional style checks, including Rubocop style checks."
```
`switch`'s first two arguments are already being specified for short and long options . eg: 
`switch   "-s", "--strict"`

GauthamGoli replied:
May be we can separate them out by checking if they start with `-` and doesn't contain any  ` `

MikeMcQuaid commented on Library/Homebrew/brew.rb:
```diff
@@ -82,7 +82,7 @@
   # - no arguments are passed
   # - if cmd is Cask, let Cask handle the help command instead
-  if (empty_argv || help_flag) && cmd != "cask"
+  if (empty_argv || help_flag) && cmd != "cask" && !internal_dev_cmd
```
Also CC @reitermarkus as a note that it'd be good to get Cask's help behaving similar to the rest of Homebrew's.

GauthamGoli commented:
I've been caught up at work.. will get back working on this in 3 days.

GauthamGoli commented:
@MikeMcQuaid From what I see, `OptionParser` is very well suited for generating `brew <command> --help`

Man pages are formatted differently, we cannot directly use the help text as-is, but it CAN be used by writing some extra code. I looked at the man pages of `git` which also has sub-commands like `brew` 

Before proceeding, I have a question to clarify on help text rephrasing.
 

- `brew audit --help`  generated without `OptionParser`

```
brew audit [--strict] [--fix] [--online] [--new-formula] [--display-cop-names] [--display-filename] [--only=method|--except=method] [--only-cops=cops|--except-cops=cops] [formulae]:
    Check formulae for Homebrew coding style violations. This should be
    run before submitting a new formula.

    If no formulae are provided, all of them are checked.

    If --strict is passed, additional checks are run, including RuboCop
    style checks.
```
(snipped)
 - `brew audit --help` generated with `OptionParser`

```
brew audit [options] [<formulae>]

        Check <formulae> for Homebrew coding style violations. This should be
        run before submitting a new formula.

        If no <formulae> are provided, all of them are checked.

        --strict                     Run additional style checks, including Rubocop style checks.
```

Since `OptionParser` displays the description adjacent to the option, I rephrased it for brevity from 
```
If --strict is passed, additional checks are run, including RuboCop style checks.
``` 
to 
```
Run additional style checks, including Rubocop style checks.
```

Do we want to keep the current help text as-is , or rephrase it to latter ?

MikeMcQuaid commented:
This doesn't look bad but would be nice to have some of the nicer formatting tweaks:

<img width="1221" alt="screenshot 2018-08-27 at 10 42 59" src="https://user-images.githubusercontent.com/125011/44653203-f9e5ec00-a9e5-11e8-9743-9a7f48017cd8.png">

On a narrower terminal the existing version wraps a bit nicer:

<img width="560" alt="screenshot 2018-08-27 at 10 43 52" src="https://user-images.githubusercontent.com/125011/44653270-2c8fe480-a9e6-11e8-9e46-3302f1d41253.png">

GauthamGoli commented:
I was facing formatting issues and had given up temporarily. I will give it a try again now to get the formatting right.

MikeMcQuaid commented:
@GauthamGoli Thanks! We could always write our own generator if that'd be better.

GauthamGoli commented:
@MikeMcQuaid Can you please generate manpage and let me know what you think?

MikeMcQuaid commented on Library/Homebrew/utils/formatter.rb:
```diff
@@ -31,6 +31,10 @@ def error(string, label: nil)
   end
 
+  def wrap(s, width = 172, indentation = 0)
+    " "*indentation + s.gsub(/(.{1,#{width}})(\s+|\Z)/, "\\1\n").gsub("\n", "\n"+" "*indentation) + "\n"
```
This is hard to follow. Could you use intermediate variables and/or string interpolation and put each `.gsub` on a new line?

MikeMcQuaid commented on Library/Homebrew/cli_parser.rb:
```diff
@@ -16,14 +16,28 @@ def initialize(&block)
+      end
+
+      def wrap_option_desc(desc)
+        Formatter.wrap(desc, @desc_line_length).split("\n")
```
What happens without these being wrapped? Is it the manpage/markdown/something else that looks weird?

MikeMcQuaid replied:
@GauthamGoli I see, thanks 👍

MikeMcQuaid commented on Library/Homebrew/cli_parser.rb:
```diff
@@ -34,9 +48,18 @@ def switch(*names, description: nil, env: nil, required_for: nil, depends_on: ni
         enable_switch(*names) if !env.nil? && !ENV["HOMEBREW_#{env.to_s.upcase}"].nil?
       end
 
+      def banner(text)
```
Is the banner the help text? If so, a name relating to that might be more obvious.

GauthamGoli replied:
Banner is a brief description of the command shown above the options. 
Ex: 
```ruby
      banner <<~EOS
         Usage: brew audit [options] [<formulae>]
         Check <formulae> for Homebrew coding style violations. This should be
         run before submitting a new formula.
         If no <formulae> are provided, all of them are checked.
      EOS
```

MikeMcQuaid replied:
@GauthamGoli Ok, I see. Maybe `usage_banner` or `help_banner` or similar?

MikeMcQuaid commented on Library/Homebrew/cli_parser.rb:
```diff
@@ -95,10 +128,10 @@ def enable_switch(*names)
-        when :verbose then [["-v", "--verbose"], :verbose]
-        when :debug   then [["-d", "--debug"], :debug]
-        when :force   then [["-f", "--force"], :force]
+        when :quiet   then [["-q", "--quiet"], :quiet, "Suppress warnings."]
```
`Suppress any warnings`

MikeMcQuaid commented on Library/Homebrew/cli_parser.rb:
```diff
@@ -95,10 +128,10 @@ def enable_switch(*names)
-        when :force   then [["-f", "--force"], :force]
+        when :quiet   then [["-q", "--quiet"], :quiet, "Suppress warnings."]
+        when :verbose then [["-v", "--verbose"], :verbose, "Verbose mode."]
+        when :debug   then [["-d", "--debug"], :debug, "Display debug info."]
```
`Display any debugging information`

MikeMcQuaid commented on Library/Homebrew/cli_parser.rb:
```diff
@@ -95,10 +128,10 @@ def enable_switch(*names)
-        when :debug   then [["-d", "--debug"], :debug]
-        when :force   then [["-f", "--force"], :force]
+        when :quiet   then [["-q", "--quiet"], :quiet, "Suppress warnings."]
+        when :verbose then [["-v", "--verbose"], :verbose, "Verbose mode."]
```
`Make some output more verbose`

MikeMcQuaid commented on Library/Homebrew/cli_parser.rb:
```diff
@@ -95,10 +128,10 @@ def enable_switch(*names)
+        when :quiet   then [["-q", "--quiet"], :quiet, "Suppress warnings."]
+        when :verbose then [["-v", "--verbose"], :verbose, "Verbose mode."]
+        when :debug   then [["-d", "--debug"], :debug, "Display debug info."]
+        when :force   then [["-f", "--force"], :force, "Override any warnings/validations."]
```
`Override warnings and enable potentially unsafe operations`

MikeMcQuaid commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -52,22 +52,35 @@
+      EOS
+      switch      "--strict", description: "Run additional style checks, including Rubocop style checks."
+      switch      "--online", description: "Run additional slower style checks that require a network connection."
+      switch      "--new-formula", description: "Run various additional style checks to determine if a new formula is eligible for Homebrew. This should be used when creating new formula and implies --strict and --online."
```
This line (and some others) is a bit too long. Maybe stick each `description:` on a new line?

MikeMcQuaid commented:
@GauthamGoli Good work on this! I think ideally if we're writing our own output I'd just aim for the output to be as close as possible to the existing rendered manpage output i.e. do stuff like `If X is passed,` and indent/wrap similarly to the existing. Anywhere where that seems to be fair too much work: shout here and we can talk about whether it's worth it.

GauthamGoli commented:
@MikeMcQuaid The current manpage output for `brew audit` command is similar to the manpage of `git` and `nmap` (I like this format). I've noticed that a lot of other command line tools use this similar formatting in manpages instead of `If X is passed,` format. 

Do you think its worth transitioning to this new format ? If not, I'll make it as close to existing format.

GauthamGoli commented:
@MikeMcQuaid  addressed the comments. Please take a look when you get time! 

Do we plan to remove the comments in the beginning of every `command.rb` that we previously used to generate help and manpages with?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -52,22 +52,55 @@
+
+      EOS
+      switch      "--strict",            description: "Run additional style checks, "\
+                                                      "including Rubocop style checks."
```
What about:

```ruby
switch      "--strict",
  description: "Run additional style checks, including Rubocop style checks."
```

GauthamGoli replied:
In the current style, one could see all the declared options in a single column (I think its more readable this way), but with above style the description would come in between. Its just a choice and I would defer the final call on this to you @MikeMcQuaid

MikeMcQuaid replied:
I think it'd be preferable to be able to have the description on a single line, thanks 👍

MikeMcQuaid commented:
> Do we plan to remove the comments in the beginning of every `command.rb` that we previously used to generate help and manpages with?

Provided e.g. `brew audit --help` works without it: yes.

> If this code style is OK, I'll do the same for other developer commands.

The manpage style is perfect. I agree it's actually better than the current style now 😍. There will be some slight inconsistency when the previous command style is used just for non-developer commands but I think it's worth it to get this merged.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -52,22 +52,53 @@
+        run before submitting a new formula.
+
+        If no <formulae> are provided, all of them are checked.
+
```
Might be nice to make it so you don't need this trailing empty line?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -52,22 +52,53 @@
+        If no <formulae> are provided, all of them are checked.
+
+      EOS
+      switch      "--strict",
```
Sorry, terrible pedantry now but given the way `description` is I'd be tempted to say that `switch "--strict",` would look better.

GauthamGoli commented:
@MikeMcQuaid   Now manpages for all dev commands are being generated through this. Can you once take a look, and let me know your thought? Thanks. 

There are minor hiccups like `--verbose` and `--debug` appearing for every sub-command.

MikeMcQuaid commented:
<img width="663" alt="screenshot 2018-09-30 at 10 56 11" src="https://user-images.githubusercontent.com/125011/46256209-77968f00-c49f-11e8-9f19-2e8d1b132dbf.png">

@GauthamGoli I think it would be good if there was a bit more visual separation between commands either through indentation, keeping options on the same line as their first line of description or something else.


> There are minor hiccups like `--verbose` and `--debug` appearing for every sub-command.

Agreed, would be good to resolve this before merge. There could be a "global options" or similar section that apply to all commands.

GauthamGoli commented:
@MikeMcQuaid  Is the indentation fine now? I made the command synopsis to be H3 ( ###) 
(but I'm not able to get the usual underline formatting to work in the sub heading. )

MikeMcQuaid commented:
> Is the indentation fine now?

Looks great 😍.

More pedantry: 
<img width="646" alt="screenshot 2018-10-01 at 14 26 16" src="https://user-images.githubusercontent.com/125011/46291291-f665f780-c585-11e8-94f0-2da23308ad13.png">

- Ideally `[options]` would have only `options` in red (or maybe green? see `install-options` and not the `[ ]`
- Ideally `formulae` would be in `green` (and have the same `[]` fixes)
- As mentioned above: ideally `--verbose`, `--debug` wouldn't be repeated

After that: I think we're good to go. So close now! Great work again @GauthamGoli!

reitermarkus commented on Library/Homebrew/dev-cmd/prof.rb:
```diff
@@ -1,7 +1,7 @@
 #:    Run Homebrew with the Ruby profiler.
 #:    For example:
-#       brew prof readall
+#:       brew prof readall
```
Should be one fewer space here.

GauthamGoli replied:
Nice catch :D thanks

reitermarkus commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -52,22 +52,51 @@
+      comma_array "--except-cops",
+        description: "Passing `--except-cops`=<cops> will skip checking the listed RuboCop cops "\
+                     "violations. `cops` should be a comma-separated list of cop names."
+      switch :verbose
```
Is there a difference between `switch "--strict"` and `switch :verbose`? If not, we should only use either strings or symbols.

GauthamGoli replied:
Symbols indicate that its a global option, and `CLI::Parser` would pick up predefined attributes related to that option.

MikeMcQuaid commented:
> Now the command synopsis text i.e., `audit [options] [formulae]` also has formatting according to the review comments

Almost but not quite; `[options]` should be used when options are optional and `formulae` is correctly without `[]`. Otherwise: looking great 👍

MikeMcQuaid commented on Library/Homebrew/dev-cmd/man.rb:
```diff
@@ -118,6 +147,10 @@ def convert_man_page(markup, target)
       ronn_output = ronn.read
       ronn_output.gsub!(%r{</var>`(?=[.!?,;:]?\s)}, "").gsub!(%r{</?var>}, "`") if format_flag == "--markdown"
+      unless format_flag == "--markdown"
+        ronn_output = ronn_output.gsub(/<code>/, "\\fB").gsub(%r{</code>}, "\\fR")
```
Put each `.gsub` on a new line and ideally document what each one is doing.

GauthamGoli replied:
I've changed it so its self documented.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/man.rb:
```diff
@@ -136,4 +169,56 @@ def target_path_to_format(target)
+        cmd_parser = Homebrew.send(cmd_arg_parser(cmd_path))
+        man_page_lines << cmd_manpage_lines(cmd_parser).join
+      rescue NoMethodError
+        man_page_lines << path_glob_commands(cmd_path.to_s)[0]
```
`.first`

GauthamGoli commented:
@MikeMcQuaid What about now? If all is fine now, I'll resolve the conflicts

GauthamGoli commented:
@MikeMcQuaid any idea why the build is failing?

MikeMcQuaid commented:
@GauthamGoli Fixing in https://github.com/Homebrew/homebrew-test-bot/pull/201 then will rerun.

MikeMcQuaid commented:
Great work again @GauthamGoli 🎉
